Add more bulk operations to Beta History

davelopez opened the pull request:
Fixes #13828

New operations supported by `api/histories/{history_id}/contents/bulk` in the API:
- [x] Change datatype (temporarily disabled)
  - [ ]  *Needs regenerating the metadata using Celery.* #14033
- [x] Change dbkey
  - [x] Validate the `dbkey` to allow only registered (and user-defined) ones
- [x] Add multiple tags
- [x] Remove multiple tags

## TODO
- [x] Add API tests
- [x] Add operations to the client

## How to test the changes?
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] Instructions for manual testing are as follows:
  - Add / remove tags in bulk
  
    ![add-remove-tags](https://user-images.githubusercontent.com/46503462/170365714-f2a5a3e7-6235-4ee6-b3fc-0801033389bc.gif)

  - Change `dbkey` in bulk
      
    ![change_dbkey](https://user-images.githubusercontent.com/46503462/172572648-0a3fd1ff-472b-4979-8c8a-d2a9e14c03d0.gif)



## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

davelopez committed:
Add more bulk operations to API

In particular:
- Change datatype
- Change dbkey
- Add tag
- Remove tag

davelopez committed:
Validate `dbkey` on bulk operations

bgruening commented on lib/galaxy/managers/genomes.py:
```diff
@@ -20,6 +20,13 @@ def __init__(self, app: StructuredApp):
 
+    def is_registered_dbkey(self, dbkey: str, user: m.User) -> bool:
+        dbkeys = self.get_dbkeys(user, chrom_info=False)
+        for item in dbkeys:
```
```suggestion
        for _, item in dbkeys:
```

Might also work when its only two elements.

davelopez replied:
Nice, Danke!

davelopez committed:
Add API test for bulk tag changes

davelopez committed:
Add API test for bulk dbkey change

davelopez committed:
Add UI to change dbkey in bulk

davelopez committed:
Allow adding or removing multiple tags in bulk

davelopez committed:
Add UI for setting tags in bulk

davelopez committed:
Add UI for removing tags in bulk

davelopez committed:
Reset to default unknown `?` genome on open modal

davelopez committed:
Disable modal OK button until tags are input

davelopez committed:
Style hack to minimize dropdown clipping

The modal dialog is too short and most of the dropdown selector gets clipped otherwise.

Better ideas of how to fix it are very welcome.

davelopez committed:
Display also general error

After failed bulk operation

davelopez committed:
Add dropdown dividers to contents operations menu

davelopez committed:
Add UI for setting datatype in bulk

davelopez committed:
Temporarily disable change_datatype bulk operation

Until metadata can be regenerated in Celery

mvdbeek commented on client/src/components/History/CurrentHistory/HistoryOperations/SelectionOperations.vue:
```diff
@@ -42,6 +43,19 @@
             </b-dropdown-item>
+            <b-dropdown-divider />
+            <b-dropdown-item v-b-modal:change-dbkey-of-selected-content data-description="change reference gnome">
+                <span v-localize>Change reference Gnome</span>
```
```suggestion
                <span v-localize>Change reference Genome</span>
```
🤣

mvdbeek replied:
I'm also gonna say this shouldn't be genome centered. the `dbkey` can refer to anything.

davelopez replied:
:laughing: :+1:

davelopez replied:
oh sorry, the previous comment didn't show up until I committed the suggestion.
Should we use `Change dbkey` or `Change database key`?

mvdbeek replied:
<img width="325" alt="Screenshot 2022-05-26 at 13 26 17" src="https://user-images.githubusercontent.com/6804901/170479090-d29fb290-5176-4d86-a2b9-cc75dfb61da6.png">
is how the single-item menu looks like ...

Can we do `Database/Build` for now and defer any decision to rename this for later ?

mvdbeek commented on client/src/components/History/CurrentHistory/HistoryOperations/SelectionOperations.vue:
```diff
@@ -60,6 +74,57 @@
+            title-tag="h2"
+            @ok="changeDbkeyOfSelected">
+            <p v-localize>Select a new reference genome for {{ numSelected }} items:</p>
+            <GenomeProvider v-slot="{ item: genomes, loading: loadingGenomes }">
```
```suggestion
            <DbKeyProvider v-slot="{ item: genomes, loading: loadingGenomes }">
```
and all the changes that follow from this ...

davelopez replied:
I can't find an existing `DbKeyProvider`. Do you mean renaming the existing `GenomeProvider` to `DbkeyProvider` and all the existing references?

mvdbeek replied:
Yes, it shouldn't have been named `GenomeProvider`.

davelopez replied:
Ohh I see, I will rename the provider and all its references in a follow-up PR, just created #13964 so I don't forget to do so :+1:

davelopez committed:
Use correct term for dbkey

Use `Database/Build` instead of `Genome`

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>
